torch/csrc/jit/codegen/cuda/transform_view.cpp

```diff
@@ -0,0 +1,711 @@
+        avt_->addSplitTransform(split_factor_, is_last_axis_rfactor);
+      } else {
+        avt_->addKeepTransform(is_last_axis_rfactor);
+      }
+    }
+
+   private:
+    const size_t split_factor_ = 0;
+  };
+
+  //! A utility class to merge any adjacent size-1 dimensions
+  class MergeAdjacentSingletonDimensions : AnalyzeMergeInterface {
+   public:
+    static void process(AnalyzeViewTransformation* avt, ViewIndexState& state) {
+      MergeAdjacentSingletonDimensions asdm(avt, state);
+      asdm.handle(
+          true /* is_index_merge_rhs */, true /* is_last_axis_rfactor */);
+    }
+
+   private:
+    MergeAdjacentSingletonDimensions(
+        AnalyzeViewTransformation* avt,
+        ViewIndexState& state)
+        : AnalyzeMergeInterface(avt, state) {}
+
+    size_t getCurrentAxisPosition() const override {
+      return state_.original_view_index + num_merge_axes_;
+    }
+
+    bool isMergeInvariantValid() const override {
+      return avt_->original_view_[getCurrentAxisPosition()] == kSingletonAxis;
+    }
+
+    void updateViewIndexState() override {
+      ++state_.original_view_index;
+      ++state_.transform_view_index;
+    }
+
+    void epilogue(bool is_last_axis_rfactor) override {}
+  };
+
+  //! Find the broadcast, merge and split operations necessary
+  //! to transform the original view into the new view
+  void findTransformation() {
+    // The original and new view are processed from left to right.
+    // old_view_index and new_view_index track the current position in each
+    // view respectively.
+    // kRfactor - Is the last iterDomain already in the rfactor domain?
+    while (state_.new_view_index < new_view_.size() &&
+           state_.original_view_index < original_view_.size()) {
+      const auto kCurrentSize = transform_view_[state_.transform_view_index];
+      auto is_last_axis_rfactor = transform_view_[state_.original_view_index] !=
+          original_view_[state_.original_view_index];
+
+      if (kCurrentSize == kEmptyAxis) {
+        // If current size in transform view is 0, then it was already handled
+        // and should be skipped.
+        ++state_.transform_view_index;
+      } else if (kCurrentSize == new_view_[state_.new_view_index]) {
+        addKeepTransform(is_last_axis_rfactor);
+      } else if (new_view_[state_.new_view_index] == kSingletonAxis) {
+        addBroadcastTransform();
+      } else {
+        MergeThenSplit::process(
+            this,
+            state_,
+            kCurrentSize,
+            new_view_[state_.new_view_index],
+            is_last_axis_rfactor);
+      }
+    }
+
+    MergeAdjacentSingletonDimensions::process(this, state_);
+
+    // Skip any root domains that were merged for any splits with remainder
+    // OR any singleton axes
+    while (state_.transform_view_index < transform_view_.size() &&
+           transform_view_[state_.transform_view_index] <= kSingletonAxis) {
+      ++state_.transform_view_index;
+    }
```

Comment:
Is this part necessary? At this point and later, it seems we don't use `transform_view_index` anymroe.

Reply:
It isn't necessary to complete the transformation, but `transform_view_index ==  transform_view_.size()` is an invariant that I check for in `validate`.